### PR TITLE
Add rest api file manager

### PR DESCRIPTION
**Pull request by cn-kali-team**
When I call brute force password cracking on the RPC interface, I need to specify a wordlist file, I had to use other methods to upload the wordlist file to the server to use it, so I added the file manager function to the rest api, and it would be easy to manage the files on the server.

- It will create a rest_files directory in the MSF configuration folder
```
➜  .msf4 pwd
/home/kali-team/.msf4
➜  .msf4 ls --icon never
database.yml  db  history  local  logos  logs  loot  modules  msf-ws-cert.pem  msf-ws-key.pem  msf-ws.pid  plugins  rest_files  store
➜  .msf4 
```
### download file
```
➜  ~ curl -k -X GET "https://api.kali-team.cn:5443/api/v1/files/file?path=A.txt" -H "accept: application/json" -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081"
A
```
### upload file
- key name for uploaded file is `file`
```
➜  ~ curl -k -F "path=/password/TOP1000.txt" -F "file=@TOP1000.txt" https://api.kali-team.cn:5443/api/v1/files/file -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081"
{"data":{"path":"/home/kali-team/.msf4/rest_files/password/TOP1000.txt"}}% 
➜  ~ cd .msf4/rest_files 
➜  rest_files ls
 A.jpg  A.txt   password  pic
➜  rest_files cd password 
➜  password ls
TOP1000.txt
➜  password 
```
### rename file
- rename B.txt to A.txt
```
➜  rest_files ls
  A.jpg   B.txt   password   pic
➜  rest_files curl -k -X PUT "https://api.kali-team.cn:5443/api/v1/files/file" -H "accept: application/json" -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081" -H "Content-Type: application/json" -d "{ \"path\": \"B.txt\", \"new_path\": \"A.txt\"}"
{"data":{"path":"/home/kali-team/.msf4/rest_files/A.txt"}}%                                                                                                                   ➜  rest_files ls
  A.jpg   A.txt   password   pic
➜  rest_files 
```
### delete file
- remove A.txt
```
➜  rest_files ls --icon never
A.jpg  A.txt  password  pic
➜  rest_files curl -k -X DELETE "https://api.kali-team.cn:5443/api/v1/files/file" -H "accept: application/json" -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081" -H "Content-Type: application/json" -d "{ \"path\": \"A.txt\"}"
{"data":{"path":"/home/kali-team/.msf4/rest_files/A.txt"}}%                                                                                                                   ➜  rest_files ls
  A.jpg    password   pic
➜  rest_files 
```

### list
```
curl -X GET "https://api.kali-team.cn:5443/api/v1/files/dir" -H "accept: application/json" -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081"
```
- response
``` json
{
  "data": [
    {
      "name": "password",
      "type": "directory",
      "size": "4096",
      "last_modified": "2020-06-02T16:56:26.185+08:00"
    },
    {
      "name": "pic",
      "type": "directory",
      "size": "4096",
      "last_modified": "2020-06-02T16:10:21.978+08:00"
    },
    {
      "name": "A.jpg",
      "type": "file",
      "size": "3642",
      "last_modified": "2020-06-02T15:59:00.698+08:00"
    }
  ]
}
```
### Create a directory
```
curl -X POST "https://api.kali-team.cn:5443/api/v1/files/dir" -H "accept: application/json" -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081" -H "Content-Type: application/json" -d "{ \"path\": \"new\"}"
```
- response
```
{
  "data": {
    "path": "/home/kali-team/.msf4/rest_files/new"
  }
}
```
### Rename directory
```
curl -X PUT "https://api.kali-team.cn:5443/api/v1/files/dir" -H "accept: application/json" -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081" -H "Content-Type: application/json" -d "{ \"path\": \"new\", \"new_path\": \"new_old\"}"
```
- response
```
{
  "data": {
    "path": "/home/kali-team/.msf4/rest_files/new_old"
  }
}
```
### delete directory
```
curl -X DELETE "https://api.kali-team.cn:5443/api/v1/files/dir" -H "accept: application/json" -H "Authorization: Bearer 15f96b3959677f6e45c2efaa1b42ae46cdcac96a2c0e41d3e2033d001af534d10842f40fe58ee081" -H "Content-Type: application/json" -d "{ \"path\": \"new_old\"}"
```
- response
```
{
  "data": {
    "path": "/home/kali-team/.msf4/rest_files/new_old"
  }
}
```

### CODE REVIEW

File: lib/msf/core/web_services/servlet/file_servlet.rb [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    next unless file != '.' && file != '..'
+
+    tbl << row
+  end
+  return tbl
+end
+
+def safe_expand_path?(path)
+  current_directory = File.expand_path(Msf::Config.rest_files_directory)
+  tested_path = File.expand_path(path)
```
Comment (smcintyre-r7) [inline lib/msf/core/web_services/servlet/file_servlet.rb:197]: `File.expand_path` does not append a path separator character to the end to ensure that we're limiting the operations to the expected path.

This should be:
```suggestion
  current_directory = File.expand_path(Msf::Config.rest_files_directory) + File::SEPARATOR
  tested_path = File.expand_path(path) + File::SEPARATOR
```

File: lib/msf/core/web_services/servlet/file_servlet.rb [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        result = { path: search_file_paths }
+      else
+        result = { message: 'Invalid parameter' }
+      end
+      set_json_data_response(response: result)
+    }
+  end
+end
+
+def list_local_path(path)
```
Comment (adfoster-r7) [inline lib/msf/core/web_services/servlet/file_servlet.rb:219]: It looks like these functions are now global, it'd be great to move them into the servlet if possible :+1:

**Review by adfoster-r7 — CHANGES_REQUESTED** [review]

File: lib/msf/core/web_services/servlet/file_servlet.rb (hunk 2 above)
Comment (cn-kali-team) [reply lib/msf/core/web_services/servlet/file_servlet.rb:219]: I am a beginner, what does servlet mean, what should I do, thank you!

Comment (cn-kali-team) [reply lib/msf/core/web_services/servlet/file_servlet.rb:219]: > It looks like these functions are global, it'd be great to move them into the servlet if possible

I am a beginner, what does servlet mean, what should I do, thank you!

**Comment by adfoster-r7** [comment]
Thanks for the pull request @cn-kali-team! It looks like this pull request is making changes to the public API of Metasploit. We try to ensure our public APIs can be supported for many years to come - and as a result we need to be certain that a chosen interface and implementation can fulfil all of the required use cases that our users may have.

In this case we're not sure if Metasploit should implement the functionality of supporting arbitrary file uploads. As this pull request is aimed at supporting a not-yet-released Metasploit frontend we've taken the decision to close this pull request for now. In terms of next steps, I liked your previous idea of adding these new capabilities to your own django middleware for now - which should allow you to progress with the new project that you've been working on :+1:

Thanks again @cn-kali-team

**Pull request closed without merging**